cpp/src/arrow/buffer-test.cc

```diff
@@ -52,6 +52,22 @@ TEST(TestBuffer, FromStdString) {
   ASSERT_EQ(static_cast<int64_t>(val.size()), buf.size());
 }
 
+TEST(TestBuffer, FromStdStringWithMemory) {
+  std::string* val = new std::string("hello, world");
+  std::shared_ptr<Buffer> buf;
+  std::string expected("hello, world");
+
+  ASSERT_OK(Buffer::FromString(*val, default_memory_pool(), &buf));
+
+  ASSERT_EQ(0, memcmp(buf->data(), val->c_str(), val->size()));
+  ASSERT_EQ(static_cast<int64_t>(val->size()), buf->size());
+
+  delete val;
```

Comment:
Note we do not use `new` and `delete` in this codebase except in rare situations (we have to use `new` to initialize `std::unique_ptr` because we do not yet have `std::make_unique`).